File: cxx/connection/Connection.cpp
/**
 * @file Connection.cpp
 * @author Ronald T. Fernandez
 * @version 1.0
 */

#include "../utils/Utils.h"

#include "Connection.h"

#include <iostream>

namespace connection {

std::shared_ptr<Connection> Connection::_instance = nullptr;
const std::string Connection::CONF_NAME = "conf/server.cfg";
bool Connection::isReferenced = false;

Connection::Connection() : orb(nullptr), poa(nullptr), properties() {
	init();
}

void Connection::close() {
	if (orb != NULL) {
		try {
			orb->destroy();
		} catch (...) {}
	}
}


void Connection::init() {
	// Get the properties
	std::map<std::string, std::string> properties = std::map<std::string, std::string>();
	utils::Utils::parseFile(CONF_NAME, properties);

	char* key = const_cast<char*>(std::string("-ORBInitRef").c_str());
	std::string valueStr = std::string("NameService=corbaname::" + properties.at("org.omg.CORBA.ORBInitialHost") + ":" + properties.at("org.omg.CORBA.ORBInitialPort")).c_str();
	char *value = (char*)valueStr.c_str();
	char* args[] = { key, value};

	std::cout << "****************** PROPERTIES *******************" << std::endl;
	for (std::map<std::string, std::string>::iterator it = properties.begin(); it != properties.end(); ++it) {
		std::cout << "* " << it->first << " : " << it->second << std::endl;
	}
	std::cout << "*************************************************" << std::endl;
	std::cout << "ARGS: " << args[0] << " " << args[1] << std::endl;
	std::cout << "*************************************************" << std::endl;

	// Initialize the ORB
	int argc=2;
	orb = CORBA::ORB_init(argc, args);

	// Obtain the root context of the name service
	referenceObject();

	// Initialize and activate the POA Manager
	PortableServer::POAManager_var pman = poa->the_POAManager();
	pman->activate();

}

void Connection::referenceObject() {
	if (!isReferenced) {
		// Obtain the root context of the name service
		poa = PortableServer::POA::_narrow(orb->resolve_initial_references("RootPOA"));
		isReferenced = true;
	}
}

std::shared_ptr<Connection> Connection::getInstance() {
	// Singleton implementation
	if (_instance == nullptr)
		_instance = std::shared_ptr<Connection>(new Connection());
	return _instance;
}

CORBA::Object_ptr Connection::getClientObject(std::string contextName, std::string componentName, std::string objectType) {
	CosNaming::NamingContext_var rootContext;

	// Get the root context
	try {
		CORBA::Object_var obj;
		obj = orb->resolve_initial_references("NameService");

		rootContext = CosNaming::NamingContext::_narrow(obj);
		if (CORBA::is_nil(rootContext)) {
			std::cerr << "Error getting the root context" << std::endl;
			return CORBA::Object::_nil();
		}
	} catch (CORBA::ORB::InvalidName& ex) {
		std::cerr << "Service does not exist" << std::endl;
		return CORBA::Object::_nil();
	}

	// Get the object given the rootContext, the context name and the componentName (object reference name) and type of object
	CosNaming::Name name;
	name.length(2);
	name[0].id = componentName.c_str();
	name[0].kind = contextName.c_str();
	name[1].id = objectType.c_str();
	name[1].kind = "Object";

	// Return the reference in case it was found or show an otherwise
	try {
		return rootContext->resolve(name);
	} catch (CosNaming::NamingContext::NotFound& e) {
		std::cerr << "Context not found" << std::endl;
	} catch (CORBA::COMM_FAILURE& e) {
		std::cerr << "Error contacting to the naming service" << std::endl;
	} catch (CORBA::SystemException& e) {
		std::cerr << "A system exception has occurred" << std::endl;
	}

	// Return a null reference in case it is not found
	return CORBA::Object::_nil();
}

void Connection::bindObjectToName(CORBA::Object_ptr objref, std::string contextName, std::string componentName, std::string objectType) {
	CosNaming::NamingContext_var namingContext;
	CosNaming::NamingContext_var subContext;

	// Get the root context
	try {
		CORBA::Object_var obj = orb->resolve_initial_references("NameService");
		namingContext = CosNaming::NamingContext::_narrow(obj);

		if (CORBA::is_nil(namingContext)) {
			std::cerr << "Failed to narrow the root naming context." << std::endl;
			throw std::exception();
		}
	} catch (CORBA::ORB::InvalidName& ex) {
		std::cerr << "Service required is not valid" << std::endl;
		throw std::exception();
	}

	// Sets the context and put there the object of objectType providing a componentName
	try {
		CosNaming::Name nameComponent;
		nameComponent.length(1);
		nameComponent[0].id = componentName.c_str();
		nameComponent[0].kind = contextName.c_str();

		try {
			// Bind the context
			subContext = namingContext->bind_new_context(nameComponent);
		} catch (CosNaming::NamingContext::AlreadyBound& ex) {
			CORBA::Object_var obj;
			obj = namingContext->resolve(nameComponent);
			subContext = CosNaming::NamingContext::_narrow(obj);
			if (CORBA::is_nil(subContext)) {
				std::cerr << "Failed to narrow naming context." << std::endl;
				throw std::exception();
			}
		}

		// Set the object name and type
		CosNaming::Name objectName;
		objectName.length(1);
		objectName[0].id = (const char*) objectType.c_str();
		objectName[0].kind = (const char*) "Object";

		try {
			// Link the object with the context and the object name and type
			subContext->bind(objectName, objref);
		} catch (CosNaming::NamingContext::AlreadyBound& ex) {
			subContext->rebind(objectName, objref);
		}

	} catch (CORBA::COMM_FAILURE& ex) {
		std::cerr << "Unable to access to the naming service" << std::endl;
		throw std::exception();
	} catch (CORBA::SystemException&) {
		std::cerr << "A system exception has occurred" << std::endl;
		throw std::exception();
	}
}

CORBA::Object_ptr Connection::activateServant(PortableServer::ServantBase* obj) {
	// Activate the CORBA object
	poa->activate_object(obj);
	return poa->servant_to_reference(obj);
}

void Connection::deactivateServant(PortableServer::ServantBase* obj) {
	// Deactivate the CORBA object
	poa->deactivate_object(*poa->servant_to_id(obj));
}

void Connection::runServer() {
	// Start a CORBA server to receive incoming requests
	orb->run();
}
};

File: java/connection/Connection.java
/**
 * @file AccountDelegate.java
 * @author Ronald T. Fernandez
 * @version 1.0
 */
package connection;

import java.nio.channels.AlreadyBoundException;
import java.util.Properties;

import org.omg.CORBA.COMM_FAILURE;
import org.omg.CORBA.ORB;
import org.omg.CORBA.Object;
import org.omg.CORBA.SystemException;
import org.omg.CORBA.ORBPackage.InvalidName;
import org.omg.CosNaming.NameComponent;
import org.omg.CosNaming.NamingContext;
import org.omg.CosNaming.NamingContextHelper;
import org.omg.CosNaming.NamingContextPackage.AlreadyBound;
import org.omg.CosNaming.NamingContextPackage.NotFound;
import org.omg.PortableServer.POA;
import org.omg.PortableServer.POAHelper;
import org.omg.PortableServer.POAManager;
import org.omg.PortableServer.Servant;
import org.omg.PortableServer.POAPackage.ServantAlreadyActive;
import org.omg.PortableServer.POAPackage.ServantNotActive;
import org.omg.PortableServer.POAPackage.WrongPolicy;

import utils.Utils;

public class Connection {

	/**
	 * ORB reference
	 */
	private ORB orb = null;
	
	/**
	 * Connection configuration file name
	 */
	private static final String CONF_NAME = "conf/server.cfg";
	
	/**
	 * Indicates whether the reference of the root context of the name service has been obtained
	 */
	private static boolean isReferenced = false;
	
	/**
	 * Connection configuration properties
	 */
	private Properties properties;
	
	/**
	 * POA reference
	 */
	private POA poa;
	
	/**
	 * Instance needed to implement the singleton pattern
	 */
	private static Connection _instance = null;
	
	/**
	 * Get an instance of the class - Singleton pattern
	 * @return A reference to the connection
	 * @throws Exception Throws an exception in case of any problem
	 */
	public static Connection getInstance() throws Exception {
		// Singleton implementation
		if (_instance == null) {
			_instance = new Connection();
		}
		return _instance;
	}
	
	/**
	 * Class constructor
	 * @throws Exception Throws an exception in case of any problem
	 */
	private Connection() throws Exception {
		init();
	}
	
	/**
	 * Initialize the attributes of the instance
	 * @throws Exception Throws an exception in case of any problem
	 */
	private void init() throws Exception {
		// Get the properties
		properties = Utils.readProperties(CONF_NAME);
		String args[] = { "-ORBInitRef", "NameService=corbaname::" + 
				properties.getProperty("org.omg.CORBA.ORBInitialHost") + ":" + 
				properties.getProperty("org.omg.CORBA.ORBInitialPort")};
		
		System.out.println("****************** PROPERTIES *******************");
		for (java.lang.Object p : properties.keySet())
			System.out.println("* " + p.toString() + " : " + properties.getProperty(p.toString()));
		System.out.println("*************************************************");
		System.out.println("ARGS: " + args[0] + " " + args[1]);
		System.out.println("*************************************************");
		
		// Intialize the ORB
		orb = ORB.init(args, properties);
		
		// Obtain the root context of the name service
		referenceObject();
		
		// Initialize and activate the POA Manager
		POAManager pman = poa.the_POAManager();
		pman.activate();
	}
	
	/**
	 * Obtain the reference of the root context of the name service
	 * @throws InvalidName	Throws this exception in case the name is not valid
	 */
	private void referenceObject() throws InvalidName {
		if (!isReferenced) {
			// Obtain the root context of the name service
			poa = POAHelper.narrow(orb.resolve_initial_references("RootPOA"));
			isReferenced = true;
		}
	}
	
	/**
	 * Resolve the CORBA naming service and get the referenced object
	 * @param contextName	Name of the context
	 * @param componentName	Name of the component
	 * @param objectType	Type of object
	 * @return Referenced object (published in the CORBA naming service)
	 * @throws Exception Throw an exception in case a generic problem occurs
	 */
	public Object getClientObject(String contextName, String componentName, String objectType) throws Exception {
		// Get the root context
	    Object objRef = orb.resolve_initial_references("NameService");
	    NamingContext namingContext = NamingContextHelper.narrow(objRef);
	    
	    // Get the object given the rootContext, the context name and the componentName (object reference name) and type of object
	    NameComponent name[] = new NameComponent[] {
	      new NameComponent(componentName, contextName),
	      new NameComponent(objectType, "Object")
	    };
	    
	    // Return the reference in case it was found or throw an exception otherwise
	    Object object = namingContext.resolve(name);
	    if (object == null) {
	      throw new Exception("Reference is null");
	    }
	    return object;
	}
	
	/**
	 * Publish the object in the CORBA naming service
	 * @param objRef		Referenced object (published in the CORBA naming service)
	 * @param contextName	Name of the context
	 * @param componentName	Name of the component
	 * @param objectType	Type of object
	 * @throws Exception Throw an exception in case a generic problem occurs
	 */
	public void bindObjectToName(Object objRef, String contextName, String componentName, String objectType) throws Exception {
		NamingContext namingContext;
		NamingContext subContext;
		
		// Get the root context
		try {
			Object object = orb.resolve_initial_references("NameService");
			namingContext = NamingContextHelper.narrow(object);
			if (namingContext == null)
				throw new Exception("A problem occurred when narrowing the naming context");
		} catch (Exception e) {
			throw new Exception("Name service does not exist");
		}
		
		// Sets the context and put there the object of objectType providing a componentName
		try {
			NameComponent nameComponent[] = new NameComponent[] {
					new NameComponent(componentName, contextName)
			};
			try {
				// Bind the context
				subContext = namingContext.bind_new_context(nameComponent);
			} catch (NotFound nf) {
				throw new Exception("Cannot bind new context");
			} catch (AlreadyBoundException abe) {
				//	throw new Exception("Context is already bound");
				Object object = namingContext.resolve(nameComponent);
				if ((subContext = NamingContextHelper.narrow(object)) == null)
					throw new Exception("Cannot rebind the naming context");
			}
			
			// Set the object name and type
			NameComponent objectName[] = new NameComponent[] {
					new NameComponent(objectType, "Object")
			};
			try {
				// Link the object with the context and the object name and type
				subContext.bind(objectName, objRef);
			} catch (AlreadyBound ab) {
				// throw new Exception("Object already bound to the subcontext");
				subContext.rebind(objectName, objRef);
			}
		} catch (COMM_FAILURE cf) {
			throw new Exception("A problem has occurred when contacting the naming service");
		} catch (SystemException se) {
			throw new Exception("A problem has occurred whe using the naming service");
		}
		System.out.println("Object of type " + objectType + " bound in componentName " + componentName + " and context " + contextName);
	}	
	
	/**
	 * Activate CORBA object
	 * @param obj	CORBA object to activate
	 * @return	Activated object
	 * @throws Exception Throw an exception in case a generic problem occurs
	 */
	public Object activateServant(Servant obj) throws Exception {
		// Activate the CORBA object
		poa.activate_object(obj);
		return poa.servant_to_reference(obj);
	}
	
	/**
	 * Starts a server to receive incoming CORBA requests
	 */
	public void runServer() {
		orb.run();
	}
	
	/**
	 * Close the connection
	 */
	public void close() {
		if (orb != null) {
			orb.destroy();
		}
	}
}
